akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -636,29 +648,40 @@ import scala.collection.JavaConverters._
   override def createLogic(attr: Attributes) = new GraphStageLogic(shape) with InHandler {
     setHandler(in, this)
 
-    override def onPush(): Unit = externalCallback(ActorSubscriberMessage.OnNext(grab(in)))
-    override def onUpstreamFinish(): Unit = externalCallback(ActorSubscriberMessage.OnComplete)
-    override def onUpstreamFailure(ex: Throwable): Unit = externalCallback(ActorSubscriberMessage.OnError(ex))
-
-    @tailrec
-    private def setCallback(callback: Command ⇒ Unit): Unit =
-      status.get match {
-        case Uninitialized ⇒
-          if (!status.compareAndSet(Uninitialized, /* Materialized */ getAsyncCallback[Command](callback)))
-            setCallback(callback)
-
-        case cmd: CommandScheduledBeforeMaterialization ⇒
-          if (status.compareAndSet(cmd, /* Materialized */ getAsyncCallback[Command](callback)))
-            // between those two lines a new command might have been scheduled, but that will go through the
-            // async interface, so that the ordering is still kept
-            callback(cmd.command)
-          else
-            setCallback(callback)
-
-        case m: /* Materialized */ AsyncCallback[Command @unchecked] ⇒
-          failStage(new IllegalStateException("Substream Source cannot be materialized more than once"))
+    override def onPush(): Unit = sendMessage(ActorSubscriberMessage.OnNext(grab(in)))
+    override def onUpstreamFinish(): Unit = sendMessage(ActorSubscriberMessage.OnComplete)
+    override def onUpstreamFailure(ex: Throwable): Unit = sendMessage(ActorSubscriberMessage.OnError(ex))
+
+    override def postStop(): Unit = {
+      externalCallback.cancel()
+    }
+
+    private def setCallback(callback: Command ⇒ Unit): Unit = {
+      @tailrec
+      def loop(asyncCallback: AsyncCallback[Command]): Unit = {
+        status.get match {
+          case Uninitialized ⇒
```

Comment:
I wonder if we actually need the logic here with the new async callbacks at all? (Could be solved in another ticket, though.)